### PR TITLE
Fix fuse open flag with O_TRUNC

### LINKED ISSUE #14895: alluxio fuse delete file unexpectedly in open function
**Alluxio Version:**
2.7.2

**Describe the bug**
alluxio fuse  not correctly deal with open flags O_RDONLY、O_WRONLY、O_RDWR，which will delete file unexpectely.

**To Reproduce**
the below code open file with O_WRONLY, which will not delete exist file in linux，however in alluxio fuse, it will delete the file,
we should delete the file only if user specify O_TRUNC flag.

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include<stdio.h>
int main()
{
    int fd, size;
    char s[] = "Linux Programmer!\n";
    char buffer[80];
    char txt[] = "mnt/test.txt";
    fd = open(txt, O_WRONLY|O_CREAT, 644);
    if (fd < 0) {
       printf("failed to open file, fd:%d", fd);
       return -1;
    }
    int len = sizeof(s);
   printf("len:%d\n", len);
    int ret = write(fd, s, len);
    if (ret < 0) {
       printf("failed to write data, ret:%d", ret);
       return ret;
    }
   close(fd);
}

**Expected behavior**
we should delete the file only if user specify O_TRUNC flag.

**Urgency**
middle level urgency

**Are you planning to fix it**
To avoid file was wrongly deleted, i'm planning to remove the exist file only if user specify O_TRUNC flag

### PR DESCRIPTION
**Pull request by LuQQiu**
Fixes #14895
### What changes are proposed in this pull request?
Fix Fuse open flag with O_TRUNC
Fix how Alluxio deal with Fuse open flags
### Why are the changes needed?
Otherwise we cannot deal with fuse open flags correctly
### Does this PR introduce any user facing changes?
Yeah, without O_TRUNC file will not be deleted

co-author liuyongqing

### CODE REVIEW

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 1]
```diff
@@ -339,13 +334,13 @@ private int getattrInternal(String path, FileStat stat) {
 
   @Override
   public int readdir(String path, long buff, long filter, long offset,
-      FuseFileInfo fi) {
+                     FuseFileInfo fi) {
```
Comment (ssz1997) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:337]: extra tabs

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 2]
```diff
@@ -339,13 +334,13 @@ private int getattrInternal(String path, FileStat stat) {
...
     return AlluxioFuseUtils.call(LOG, () -> readdirInternal(path, buff, filter, offset, fi),
         "Fuse.Readdir", "path=%s,buf=%s", path, buff);
   }
 
   private int readdirInternal(String path, long buff, long filter, long offset,
-      FuseFileInfo fi) {
+                              FuseFileInfo fi) {
```
Comment (ssz1997) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:343]: extra tabs

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 3]
```diff
@@ -461,36 +462,17 @@ private int readInternal(String path, ByteBuffer buf, long size, long offset, Fu
...
-                mOpenFileEntries.put(fd, is);
-              } finally {
-                mOpenReadWriteLocks.remove(fd);
-              }
-            }
-          }
+        final int flags = fi.flags.get();
+        if (AlluxioFuseOpenUtils.getOpenAction(flags) == OpenAction.READ_WRITE) {
+          LOG.error(String.format("Alluxio only supports read-only or write-only. "
+                  + "Path %s is opened with flag 0x%x for reading and writing concurrently. "
```
Comment (ssz1997) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:468]: 468-470, extra tabs

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 4]
```diff
@@ -1059,14 +999,14 @@ public void umount(boolean force) throws FuseException {
       try {
         CommonUtils.waitFor("all in progress file read/write to finish",
             () -> mCreateFileEntries.isEmpty() && mOpenFileEntries.isEmpty(),
-                WaitForOptions.defaults().setTimeoutMs(mMaxUmountWaitTime));
+            WaitForOptions.defaults().setTimeoutMs(mMaxUmountWaitTime));
```
Comment (ssz1997) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:1002]: missing tab?

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 5]
```diff
@@ -1059,14 +999,14 @@ public void umount(boolean force) throws FuseException {
...
       } catch (InterruptedException e) {
         LOG.error("Unmount {} interrupted", mMountPoint);
         Thread.currentThread().interrupt();
       } catch (TimeoutException e) {
         LOG.error("Timeout when waiting all in progress file read/write to finish "
-            + "when unmounting {}. {} fileInStream remain unclosed. "
-            + "{} fileOutStream remain unclosed.",
+                + "when unmounting {}. {} fileInStream remain unclosed. "
```
Comment (ssz1997) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:1008]: 1008, 1009 extra tabs

**Review by ssz1997 — APPROVED** [review]
Much cleaner. Thanks. A few extra/missing tabs though

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 4 above)
Comment (LuQQiu) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:1002]: this one is actually correct, thanks

**Comment by LuQQiu** [comment]
Error: 2.682 [ERROR] alluxio.master.journal.raft.RaftJournalTest.gainPrimacyAfterSuspend  Time elapsed: 32.291 s  <<< ERROR!
java.util.concurrent.TimeoutException: Timed out waiting for follower becomes leader options: WaitForOptions{interval=20, timeout=30000} last value: false
	at alluxio.util.CommonUtils.waitForResult(CommonUtils.java:384)
	at alluxio.util.CommonUtils.waitFor(CommonUtils.java:342)
https://github.com/Alluxio/alluxio/runs/5041547821?check_suite_focus=true

**Comment by LuQQiu** [comment]
Error: 6.924 [ERROR] alluxio.master.AlluxioMasterProcessTest.stopAfterStandbyTransition[1]
Error: 6.925 [ERROR]   Run 1: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
Error: 6.927 [ERROR]   Run 2: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
19:04:56.928 [INFO] 
https://github.com/Alluxio/alluxio/runs/5041547902?check_suite_focus=true

**Comment by LuQQiu** [comment]
Error: 3.754 [ERROR] alluxio.master.AlluxioMasterProcessTest.stopAfterStandbyTransition[0]
Error: 3.755 [ERROR]   Run 1: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
Error: 3.756 [ERROR]   Run 2: AlluxioMasterProcessTest.stopAfterStandbyTransition:209
00:08:03.756 [INFO] 
https://github.com/Alluxio/alluxio/runs/5044676445?check_suite_focus=true
alluxio.master.AlluxioMasterProcessTest.stopAfterStandbyTransition[0]  Time elapsed: 12.003 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at org.junit.Assert.assertFalse(Assert.java:75)
	at alluxio.master.AlluxioMasterProcessTest.stopAfterStandbyTransition(AlluxioMasterProcessTest.java:209)

**Comment by LuQQiu** [comment]
Error: 3.121 [ERROR] alluxio.client.fs.io.FileOutStreamIntegrationTest.cancelWrite[2]
Error: 3.121 [ERROR]   Run 1: FileOutStreamIntegrationTest.cancelWrite:248 expected:<0> but was:<1000>
Error: 3.122 [ERROR]   Run 2: FileOutStreamIntegrationTest.cancelWrite:248 expected:<0> but was:<2000>
https://github.com/Alluxio/alluxio/runs/5044676925?check_suite_focus=true

**Comment by LuQQiu** [comment]
Error: 1.179 [ERROR] alluxio.job.master.JobMasterIntegrationTest.stopJobWorkerTasks
Error: 1.181 [ERROR]   Run 1: JobMasterIntegrationTest.stopJobWorkerTasks:172 » Timeout Timed out waiting fo...
Error: 1.181 [ERROR]   Run 2: JobMasterIntegrationTest.stopJobWorkerTasks:172 » Timeout Timed out waiting fo...
00:08:31.181 [INFO] 
Error: 1.182 [ERROR] alluxio.job.master.JobMasterIntegrationTest.throttleJobWorkerTasks
Error: 1.182 [ERROR]   Run 1: JobMasterIntegrationTest.throttleJobWorkerTasks:197 » Timeout Timed out waitin...
Error: 1.182 [ERROR]   Run 2: JobMasterIntegrationTest.throttleJobWorkerTasks:197 » Timeout Timed out waitin...
https://github.com/Alluxio/alluxio/runs/5044676639?check_suite_focus=true

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
working on adding integration tests

**Pull request merged**